plutus-contract/src/Wallet/Emulator/Wallet.hs

```diff
@@ -477,36 +474,45 @@ handleBalanceTx utxo utx = do
         logDebug $ AddingCollateralInputsFor remainingCollFees
         addCollateral utxo remainingCollFees tx''
 
--- | Adjust the left and right balance of an unbalanced 'Tx' with the missing
--- lovelace considering the minimum lovelace per transaction output constraint
--- from the Cardano blockchain.
-adjustBalanceWithMissingLovelace
-    :: (Value, Value) -- ^ The unbalanced tx's left and right balance.
-    -> (Value, Value) -- ^ New left and right balance.
-adjustBalanceWithMissingLovelace (neg, pos) = do
-
-    -- We find the missing lovelace from the new positive balance. If
-    -- the positive balance is > 0 and < 'Ledger.minAdaTxOut',
-    -- then we adjust it to the minimum Ada.
-    let missingLovelaceFromPosValue =
-          if valueIsZeroOrHasMinAda pos
-            then 0
-            else max 0 (Ledger.minAdaTxOut - Ada.fromValue pos)
-        -- We calculate the final negative and positive balances
-        newPos = pos <> Ada.toValue missingLovelaceFromPosValue
-        newNeg = neg <> Ada.toValue missingLovelaceFromPosValue
-
-    (newNeg, newPos)
-
--- | Split value into an ada-only and an non-ada-only value, making sure each has at least minAdaTxOut.
-splitOffAdaOnlyValue :: Value -> [Value]
-splitOffAdaOnlyValue vl = if Value.isAdaOnlyValue vl || ada < Ledger.minAdaTxOut then [vl] else [Ada.toValue ada, vl <> Ada.toValue (-ada)]
-    where
-        ada = Ada.fromValue vl - Ledger.minAdaTxOut
-
-addOutput :: PaymentPubKey -> Maybe StakePubKey -> Value -> Tx -> Tx
-addOutput pk sk vl tx = tx & over Tx.outputs (++ pkos) where
-    pkos = (\v -> Tx.pubKeyTxOut v pk sk) <$> splitOffAdaOnlyValue vl
+calculateTxChanges
+    :: ( Member (Error WAPI.WalletAPIError) effs
+       )
+    => Params
+    -> Address -- ^ The address for the change output
+    -> Map.Map TxOutRef ChainIndexTxOut -- ^ The current wallet's unspent transaction outputs.
+    -> (Value, Value) -- ^ The unbalanced tx's negative and positive balance.
+    -> Eff effs ((Value, [TxIn]), (Value, [TxOut]))
+calculateTxChanges params addr utxos (neg, pos) = do
```

Comment:
Feels like it would be nice to have a property test here to verify that we always terminate.